GeosCore/fast_jx_mod.F90

```diff
@@ -4640,774 +2395,535 @@ SUBROUTINE LEGND0 (X,PL,N)
-       C_NIT      = State_Chm%Species(id_NIT)%Conc(I,J,L)
-       C_SALA     = State_Chm%Species(id_SALA)%Conc(I,J,L)
-
-       ! Scaling factor for J(NIT)
-       ! Prevent div by zero, and also set to a minimum of 0.1
-       FAC        = SAFE_DIV( C_SALA, ( C_SALA + C_NIT ), 1.0_fp )
-       FAC        = MAX( 0.1_fp, FAC )
-
-       ! Scale factors for NITs and NIT
-       JscaleNITs = Input_Opt%hvAerNIT_JNITs
-       JscaleNIT  = Input_Opt%hvAerNIT_JNIT
-
-       ! Convert reaction channel % to a fraction
-       JNITChanA  = Input_Opt%JNITChanA / 100.0_fp
-       JNITChanB  = Input_Opt%JNITChanB / 100.0_fp
-
-       ! Set the photolysis rate of NITs
-       ZPJ(L,RXN_JNITSa,I,J) = ZPJ(L,RXN_JHNO3,I,J) * JscaleNITs
-       ZPJ(L,RXN_JNITSb,I,J) = ZPJ(L,RXN_JHNO3,I,J) * JscaleNITs
-
-       ! Set the photolysis rate of NIT
-       ZPJ(L,RXN_JNITa, I,J) = ZPJ(L,RXN_JHNO3,I,J) * JscaleNIT * FAC
-       ZPJ(L,RXN_JNITb, I,J) = ZPJ(L,RXN_JHNO3,I,J) * JscaleNIT * FAC
-
-       ! NOTE: channel scaling is 1 in FJX_j2j.dat, then updated here
-       ZPJ(L,RXN_JNITSa,I,J) = ZPJ(L,RXN_JNITSa,I,J) * JNITChanA
-       ZPJ(L,RXN_JNITa, I,J) = ZPJ(L,RXN_JNITa, I,J) * JNITChanA
-       ZPJ(L,RXN_JNITSb,I,J) = ZPJ(L,RXN_JNITSb,I,J) * JNITChanB
-       ZPJ(L,RXN_JNITb, I,J) = ZPJ(L,RXN_JNITb, I,J) * JNITChanB
-    ENDIF
-
-    !========================================================================
-    ! SPECIAL TREATMENT FOR H2SO4+hv -> SO2 + 2OH
-    !
-    ! Only allow photolysis of H2SO4 when gaseous (SDE 04/11/13)
-    !========================================================================
-
-    ! Calculate if H2SO4 expected to be gaseous or aqueous
-    ! Only allow photolysis above 6 hPa
-    ! RXN_H2SO4 specifies SO4 + hv -> SO2 + OH + OH
-    ZPJ(L,RXN_H2SO4,I,J) = ZPJ(L,RXN_H2SO4,I,J) * FRAC
+    INTEGER I
+    REAL(fp)  DEN
 
-    !========================================================================
-    ! SPECIAL TREATMENT FOR O3+hv -> O+O2
-    !
-    ! [O1D]ss=J[O3]/(k[H2O]+k[N2]+k[O2])
-    ! SO, THE EFFECTIVE J-VALUE IS J*k[H2O]/(k[H2O]+k[N2]+k[O2])
-    !
-    ! We don't want to do this if strat-chem is in use, as all
-    ! the intermediate reactions are included - this would be
-    ! double-counting (SDE 04/01/13)
-    !========================================================================
+    !=================================================================
+    ! LEGND0 begins here!
+    !=================================================================
 
-    ! Need to subtract O3->O1D from rate
-    ! RXN_O3_1 specifies: O3 + hv -> O2 + O
-    ! RXN_O3_2 specifies: O3 + hv -> O2 + O(1D)
-    ZPJ(L,RXN_O3_1,I,J) = ZPJ(L,RXN_O3_1,I,J) - ZPJ(L,RXN_O3_2,I,J)
+    ! Always does PL(2) = P[1]
+    PL(1) = 1.e+0_fp
+    PL(2) = X
+    do I = 3,N
+       DEN = (I-1)
+       PL(I) = PL(I-1)*X*(2.e+0_fp-1.0/DEN) - PL(I-2)*(1.e+0_fp-1.e+0_fp/DEN)
+    enddo
 
-  END SUBROUTINE PHOTRATE_ADJ
+  END SUBROUTINE LEGND0
 !EOC
```

Comment:
This module has many changes but most of them involve (1) passing arguments to subroutines; (2) removing GEOS-Chem specific-code (which will be handled at a higher level); and (3) using the fields from State_Chm%Phot.